Better std::fmt::Debug output on Entity and BuiltEntity

It would be handy (no pun intended!) to be able to call `println!` and pass in an `Entity` (or `BuiltEntity`) and get a list of components returned.

Currently, `Entity` shows the following debug output (`println!("entity: {:?}", entity);`):

```
entity: 0v0
```

... and `BuiltEntity` does not implement `std::fmt::Debug` afaik.

What I'd like to see is something like this:

```
let e = world.spawn(("abc", 123));
println!("entity: {:?}", e);
...
[&str, i32]
```

Or perhaps less test-y and more practically:


```
let c1 = Transform::new();
let c2 = Character::new();
let c3 = Name::new("Player1");
let c4 = Controllable::new();
let e = world.spawn((c1, c2, c3, c4));
println!("entity: {:?}", e);
...
[Transform, Character, Name("Player1"}, Controllable]
```

Thanks for the guidance!

I've tried to implement this as much as I know how based on your example code, but I'm running into what I think is a lifetime mismatch? Here's a test case:

```
type EntityRefToStringFn = Box<dyn Fn(EntityRef<'_>) -> Option<String>>;
#[test]
fn test_named_components() {
    let mut component_name_fns: Vec<EntityRefToStringFn> = Vec::new();

    let str_name_fn: EntityRefToStringFn = Box::new(|e| e.get::<&str>().map(|c| c.to_string()));
    component_name_fns.push(str_name_fn);

    let mut world = World::new();
    let e1 = world.spawn(("def", 456, true));

    let names = &component_name_fns.map(|box_fn| box_fn(world.get(e1).unwrap()));
    println!("names: {:?}", &names);
}
```

And here's what the compiler says:

```
   Compiling try-hecs v0.1.0 (/Users/duane/Projects/ayanara/src/try-hecs)
error[E0599]: no method named `map` found for type `std::vec::Vec<std::boxed::Box<(dyn for<'r> std::ops::Fn(hecs::borrow::EntityRef<'r>) -> std::option::Option<std::string::String> + 'static)>>` in the current scope
  --> tests/tests.rs:46:37
   |
46 |     let names = &component_name_fns.map(|box_fn| box_fn(world.get(e1).unwrap()));
   |                                     ^^^ method not found in `std::vec::Vec<std::boxed::Box<(dyn for<'r> std::ops::Fn(hecs::borrow::EntityRef<'r>) -> std::option::Option<std::string::String> + 'static)>>`
   |
   = note: the method `map` exists but the following trait bounds were not satisfied:
           `&mut [std::boxed::Box<(dyn for<'r> std::ops::Fn(hecs::borrow::EntityRef<'r>) -> std::option::Option<std::string::String> + 'static)>] : std::iter::Iterator`
           `&mut std::vec::Vec<std::boxed::Box<(dyn for<'r> std::ops::Fn(hecs::borrow::EntityRef<'r>) -> std::option::Option<std::string::String> + 'static)>> : std::iter::Iterator`
```

What am I missing?

Ok, I think I'm closer: adding `iter()` to `component_name_fns.iter().map(...)` gets me here:

```
error[E0308]: mismatched types
  --> tests/tests.rs:46:64
   |
46 |     let names = &component_name_fns.iter().map(|box_fn| box_fn(world.get(e1).unwrap()));
   |                                                                ^^^^^^^^^^^^^^^^^^^^^^
   |                                                                |
   |                                                                expected struct `hecs::borrow::EntityRef`, found struct `hecs::borrow::Ref`
   |                                                                help: consider dereferencing the type: `*world.get(e1).unwrap()`
   |
   = note: expected struct `hecs::borrow::EntityRef<'_>`
              found struct `hecs::borrow::Ref<'_, _>`
```

--- a/examples/format.rs
+++ b/examples/format.rs
@@ -1,0 +1,35 @@
+//! One way to the contents of an entity, as you might do for debugging. A similar pattern could
+//! also be useful for serialization, or other row-oriented generic operations.
+
+fn format_entity(entity: hecs::EntityRef<'_>) -> String {
+    fn fmt<T: hecs::Component + std::fmt::Display>(entity: hecs::EntityRef<'_>) -> Option<String> {
+        Some(entity.get::<T>()?.to_string())
+    }
+
+    const FUNCTIONS: &[&dyn Fn(hecs::EntityRef<'_>) -> Option<String>] =
+        &[&fmt::<i32>, &fmt::<bool>, &fmt::<f64>];
+
+    let mut out = String::new();
+    for f in FUNCTIONS {
+        if let Some(x) = f(entity) {
+            if out.is_empty() {
+                out.push_str("[");
+            } else {
+                out.push_str(", ");
+            }
+            out.push_str(&x);
+        }
+    }
+    if out.is_empty() {
+        out.push_str(&"[]");
+    } else {
+        out.push(']');
+    }
+    out
+}
+
+fn main() {
+    let mut world = hecs::World::new();
+    let e = world.spawn((42, true));
+    println!("{}", format_entity(world.entity(e).unwrap()));
+}

